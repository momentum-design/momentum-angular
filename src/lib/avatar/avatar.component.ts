/** @component avatar */

import {
  Component,
  ElementRef,
  EventEmitter,
  Input,
  OnChanges,
  OnInit,
  Output,
  ViewChild,
} from '@angular/core';

export type AvatarSize =
  | 'xsmall'
  | 'small'
  | 'medium'
  | 'large'
  | 'xlarge'
  | 18
  | 24
  | 28
  | 36
  | 40
  | 44
  | 52
  | 56
  | 72
  | 80
  | 84;
export type AvatarType =
  | ''
  | 'active'
  | 'bot'
  | 'call'
  | 'dnd'
  | 'group'
  | 'inactive'
  | 'meeting'
  | 'ooo'
  | 'presenting'
  | 'self'
  | 'typing';

@Component({
  selector: 'md-avatar',
  template: `
    <button
      md-button
      *ngIf="clickable; else avatarTpl"
      [attr.aria-label]="ariaLabel"
      [class]="buttonClassName"
      [circle]="true"
      [removeStyle]="true"
    >
      <ng-container *ngTemplateOutlet="avatarTpl"></ng-container>
    </button>

    <ng-template #avatarTpl>
      <div
        *ngIf="clickable; else overlayTpl"
        class="md-avatar"
        [ngClass]="[
          'md-avatar--clickable',
          (type && 'md-avatar--' + type) || '',
          (size && 'md-avatar--' + size) || '',
          (theme && 'md-avatar--' + theme) || '',
          (isDecrypting && 'md-decrypting') || '',
          className || ''
        ]"
        [title]="!hideDefaultTooltip ? title : ''"
      >
        <ng-container *ngTemplateOutlet="overlayTpl"></ng-container>
      </div>
    </ng-template>

    <ng-template #overlayTpl>
      <ng-container *ngTemplateOutlet="childrenTpl"></ng-container>
      <md-loading *ngIf="type === 'typing'"></md-loading>
      <span *ngIf="failureBadge" class="md-avatar__failure-badge"></span>
      <span
        *ngIf="hasNotification"
        class="md-avatar__notification-badge"
      ></span>
    </ng-template>

    <ng-template #childrenTpl>
      <span
        *ngIf="type === 'self'; else imageTpl"
        class="md-avatar__self"
        [ngStyle]="{
          'background-color': backgroundColor,
          color: color
        }"
      >
        <md-icon
          [name]="
            size === 40 || size === 'medium'
              ? 'chat-active_18'
              : 'chat-active_16'
          "
          color="blue-50"
        ></md-icon>
      </span>
    </ng-template>

    <ng-template #imageTpl>
      <ng-container *ngIf="src && !isImageErrored; else iconTpl">
        <ng-container *ngIf="title && !isImageLoaded">
          <ng-container *ngTemplateOutlet="letterTpl"></ng-container>
        </ng-container>
        <img
          #image
          [alt]="alt"
          class="md-avatar__img"
          [ngClass]="{ 'md-avatar__img--hidden': !isImageLoaded }"
          draggable="false"
          (error)="handleImgError()"
          (load)="handleImgLoaded()"
          [src]="src"
        />
      </ng-container>
    </ng-template>

    <ng-template #iconTpl>
      <span
        *ngIf="icon; else letterTpl"
        class="md-avatar__icon"
        [ngClass]="{ 'md-avatar__icon--overview': isOverview }"
        [ngStyle]="{
          'background-color': backgroundColor,
          color: color
        }"
      >
        <md-icon [name]="icon"></md-icon>
      </span>
    </ng-template>

    <ng-template #letterTpl>
      <span
        class="md-avatar__letter"
        [ngClass]="{
          'md-decrypting': isDecrypting,
          'md-avatar__img--hidden': isImageLoaded
        }"
        [ngStyle]="{
          'background-color': backgroundColor,
          color: color
        }"
        >{{ getInitials() }}</span
      >
    </ng-template>
  `,
  host: {
    '[attr.title]': '!hideDefaultTooltip ? title : ""',
    class: 'md-avatar md-avatar--medium',
    '[class.md-decrypting]': '!clickable && isDecrypting',
  },
})
export class AvatarComponent implements OnInit, OnChanges {
  @Input() public ariaLabel: string = '';
  /** @option Image alt tag | '' */
  @Input() public alt: string = '';
  /** @option Set Avatar background color | '' */
  @Input() public backgroundColor: string = '';
  /** @option Optional css class string for button | '' */
  @Input() public buttonClassName: string = '';
  /** @option Optional css class string for Avatar component | null */
  @Input() public className: string;
  /** @option Set Avatar text color | '' */
  @Input() public color: string = '';
  /** @option Set existance of a failureBadge on the Avatar | false */
  @Input() public failureBadge: boolean = false;
  /** @option Set existance of a notification on the Avatar | false */
  @Input() public hasNotification: boolean = false;
  /** @option Set the visibility of Avatar's default tooltip | false */
  @Input() public hideDefaultTooltip: boolean = false;
  /** @option Optional icon name for the Avatar | null */
  @Input() public icon: string;
  /** @option Set if Avatar's content is decrypting | false */
  @Input() public isDecrypting: boolean = false;
  /** @option Set existance of Avatar's Overview | false */
  @Input() public isOverview: boolean = false;
  /** @option Set the size of the Avatar from one of the preconfigured options | 'medium' */
  @Input()
  get size(): AvatarSize {
    return this._size;
  }
  set size(value: AvatarSize) {
    if (!this.clickable) {
      this.elementRef.nativeElement.classList.remove(
        `md-avatar--${this._size}`
      );
      this.elementRef.nativeElement.classList.add(`md-avatar--${value}`);
    }
    this._size = value;
  }
  /** @option Optional image source for the Avatar | null */
  @Input() public src: string;
  /** @option Optional Avatar color theme | null */
  @Input()
  get theme(): string {
    return this._theme;
  }
  set theme(value: string) {
    if (!this.clickable) {
      this.elementRef.nativeElement.classList.remove(
        `md-avatar--${this._theme}`
      );
      this.elementRef.nativeElement.classList.add(`md-avatar--${value}`);
    }
    this._theme = value;
  }
  /** @option set Avatar title / user's name | null */
  @Input() public title: string;
  /** @option optional Avatar type | '' */
  @Input()
  get type(): AvatarType {
    return this._type;
  }
  set type(value: AvatarType) {
    if (!this.clickable) {
      this.elementRef.nativeElement.classList.remove(
        `md-avatar--${this._type}`
      );
      this.elementRef.nativeElement.classList.add(`md-avatar--${value}`);
    }
    this._type = value;
  }

  @Output() click: EventEmitter<any> = new EventEmitter();

  @ViewChild('image') image: ElementRef;

  private _size: AvatarSize = 'medium';
  private _type: AvatarType = '';
  private _theme: string;
  public clickable: boolean = false;
  public isImageLoaded: boolean = false;
  public isImageErrored: boolean = false;

  constructor(private elementRef: ElementRef) {}

  ngOnInit() {
    const img = this.image;
    if (img && img.nativeElement.complete) {
      this.handleImgLoaded();
    }
    this.clickable = this.click.observers.length > 0;
  }

  ngOnChanges(changes: any) {
    if (changes.src && changes.src.previousValue !== changes.src.currentValue) {
      this.handleImgChange();
    }
  }

  handleImgChange() {
    this.isImageLoaded = false;
    this.isImageErrored = false;
  }

  handleImgError() {
    this.isImageErrored = true;
  }

  handleImgLoaded() {
    this.isImageLoaded = true;
  }

  getInitials() {
    if (!this.title || !this.title.replace(/\s/g, '').length) {
      return '';
    }
    const letters = [];
    const words = this.title.trim().split(/ +/);
    const repeatTimes = Math.min(
      (this.type === 'group' && 1) || 2,
      words.length
    );
    for (let i = 0; i < repeatTimes; i++) {
      letters.push(String.fromCodePoint(words[i].codePointAt(0)));
    }
    return letters.join('');
  }
}
