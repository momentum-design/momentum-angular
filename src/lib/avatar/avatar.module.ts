import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonModule } from '../button';
import { IconModule } from '../icon';
import { LoadingModule } from '../loading';
import { AvatarComponent } from './avatar.component';

@NgModule({
  imports: [CommonModule, ButtonModule, IconModule, LoadingModule],
  declarations: [AvatarComponent],
  exports: [AvatarComponent],
})
export class AvatarModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: AvatarModule,
      providers: [],
    };
  }
}
