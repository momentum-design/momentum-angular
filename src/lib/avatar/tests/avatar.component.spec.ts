import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { Component } from '@angular/core';
import { AvatarComponent } from '../avatar.component';
import { AvatarModule } from '../avatar.module';

describe('AvatarComponent', () => {
  let testComponent: AvatarComponent;
  let fixture: ComponentFixture<AvatarComponent>;

  const testTitle = 'Test Name';
  const testSrc = 'test.png';
  const testIcon = 'handset_24';

  @Component({
    selector: 'test-app',
    template: `
      <md-avatar
        title="Test Name"
        src="test.png"
        ariaLabel="Test onClick"
        (click)="onClick()"
      ></md-avatar>
    `,
  })
  class TestAppComponent {
    count = 0;

    onClick() {
      this.count++;
    }
  }

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TestAppComponent],
      imports: [AvatarModule],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(AvatarComponent);
    testComponent = fixture.componentInstance;
  });

  it('should match snapshot', () => {
    testComponent.title = testTitle;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('sequence of image load', () => {
    testComponent.title = testTitle;
    testComponent.src = testSrc;
    fixture.detectChanges();
    const element = fixture.nativeElement;
    let img = element.querySelector('.md-avatar__img');
    expect(img.className).toContain('md-avatar__img--hidden');
    let letter = element.querySelector('.md-avatar__letter');
    expect(letter).not.toBeNull();

    testComponent.isImageLoaded = true;
    fixture.detectChanges();
    img = element.querySelector('.md-avatar__img');
    expect(img.className).not.toContain('md-avatar__img--hidden');
    letter = element.querySelector('.md-avatar__letter');
    expect(letter).toBeNull();
  });

  it('should handle isDecrypting prop', () => {
    testComponent.title = testTitle;
    testComponent.isDecrypting = true;
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.md-decrypting');
    expect(element).not.toBeNull();
  });

  it('should handle isOverview prop', () => {
    testComponent.icon = testIcon;
    testComponent.isOverview = true;
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector(
      '.md-avatar__icon--overview'
    );
    expect(element).not.toBeNull();
  });

  it('should trim title', () => {
    testComponent.title = '   Test Name   ';
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.md-avatar__letter');
    expect(element.textContent).toEqual('TN');
  });

  it('should display title for user', () => {
    testComponent.title = testTitle;
    testComponent.src = testSrc;
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.md-avatar__letter');
    expect(element.textContent).toEqual('TN');
  });

  it('should display title for group', () => {
    testComponent.title = 'Test Group';
    testComponent.src = testSrc;
    testComponent.type = 'group';
    fixture.detectChanges();
    const element = fixture.nativeElement.querySelector('.md-avatar__letter');
    expect(element.textContent).toEqual('T');
    expect(fixture.nativeElement.getAttribute('title')).toEqual('Test Group');
  });

  it('should handle dark theme', () => {
    testComponent.theme = 'dark';
    fixture.detectChanges();
    expect(fixture.nativeElement.className).toContain('md-avatar--dark');
  });

  describe('should apply respective classes for types', () => {
    it('when the type is group', () => {
      testComponent.title = 'Test Group';
      testComponent.src = testSrc;
      testComponent.type = 'group';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--group');
    });

    it('when the type is active', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.type = 'active';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--active');
    });

    it('when the type is inactive', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.type = 'inactive';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--inactive');
    });

    it('when the type is Do Not Disturb', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.type = 'dnd';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--dnd');
    });

    it('when the type is Out of Office', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.type = 'ooo';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--ooo');
    });

    it('when the type is call', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.type = 'call';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--call');
    });

    it('when the type is meeting', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.type = 'meeting';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--meeting');
    });

    it('when the type is presenting', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.type = 'presenting';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain(
        'md-avatar--presenting'
      );
    });

    it('when the type is typing', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.type = 'typing';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--typing');
      expect(
        fixture.nativeElement.querySelector('.md-loading')
      ).not.toBeNull();
    });

    it('when the type is bot', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.type = 'bot';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--bot');
    });

    describe('should for type self apply correct icon size', () => {
      it('when size is 36', () => {
        testComponent.title = testTitle;
        testComponent.src = testSrc;
        testComponent.type = 'self';
        testComponent.size = 36;
        fixture.detectChanges();
        expect(
          fixture.nativeElement.querySelector('.md-icon').className
        ).toContain('chat-active_16');
        expect(
          fixture.nativeElement.querySelector('.md-avatar__self')
        ).not.toBeNull();
      });

      it('when size is default(medium)', () => {
        testComponent.title = testTitle;
        testComponent.src = testSrc;
        testComponent.type = 'self';
        fixture.detectChanges();
        expect(
          fixture.nativeElement.querySelector('.md-icon').className
        ).toContain('chat-active_18');
        expect(
          fixture.nativeElement.querySelector('.md-avatar__self')
        ).not.toBeNull();
      });

      it('when size is 40', () => {
        testComponent.title = testTitle;
        testComponent.src = testSrc;
        testComponent.type = 'self';
        testComponent.size = 40;
        fixture.detectChanges();
        expect(
          fixture.nativeElement.querySelector('.md-icon').className
        ).toContain('chat-active_18');
        expect(
          fixture.nativeElement.querySelector('.md-avatar__self')
        ).not.toBeNull();
      });
    });

    it('when the failureBadge is true', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.failureBadge = true;
      fixture.detectChanges();
      expect(
        fixture.nativeElement.querySelector('.md-avatar__failure-badge')
      ).not.toBeNull();
    });

    it('when the hasNotification is true', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.hasNotification = true;
      fixture.detectChanges();
      expect(
        fixture.nativeElement.querySelector('.md-avatar__notification-badge')
      ).not.toBeNull();
    });
  });

  describe('should apply respective classes for size', () => {
    it('when size is 18/xsmall', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.size = 'xsmall';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--xsmall');

      testComponent.size = 18;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--18');
    });

    it('when size is 24', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.size = 24;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--24');
    });

    it('when size is 28/small', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.size = 'small';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--small');

      testComponent.size = 28;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--28');
    });

    it('when size is 36', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.size = 36;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--36');
    });

    it('when size is 40/medium(default)', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--medium');

      testComponent.size = 40;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--40');
    });

    it('when size is 44', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.size = 44;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--44');
    });

    it('when size is 52/large', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.size = 'large';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--large');

      testComponent.size = 52;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--52');
    });

    it('when size is 56', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.size = 56;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--56');
    });

    it('when size is 72', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.size = 72;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--72');
    });

    it('when size is 84', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.size = 84;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--84');
    });

    it('when size is 84/xlarge', () => {
      testComponent.title = testTitle;
      testComponent.src = testSrc;
      testComponent.size = 'xlarge';
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--xlarge');

      testComponent.size = 84;
      fixture.detectChanges();
      expect(fixture.nativeElement.className).toContain('md-avatar--84');
    });
  });

  it('should apply clickable class when there is a click observer', () => {
    const fixtureApp = TestBed.createComponent(TestAppComponent);
    fixtureApp.detectChanges();
    const button = fixtureApp.nativeElement.querySelector('button');
    const avatar = button.querySelector('.md-avatar');
    expect(avatar.className).toContain('md-avatar--clickable');
  });

  it('should not apply clickable class when there is no click observer', () => {
    testComponent.title = testTitle;
    testComponent.src = testSrc;
    fixture.detectChanges();
    expect(fixture.nativeElement.querySelector('button')).toBeNull();
    expect(
      fixture.nativeElement.querySelector('.md-avatar--clickable')
    ).toBeNull();
  });

  it('should display icon', () => {
    testComponent.title = testTitle;
    testComponent.icon = testIcon;
    fixture.detectChanges();
    expect(
      fixture.nativeElement.querySelector('.md-avatar__icon')
    ).not.toBeNull();
  });

  it('should display tooltip', () => {
    testComponent.title = testTitle;
    testComponent.hideDefaultTooltip = true;
    fixture.detectChanges();
    expect(fixture.nativeElement.getAttribute('title')).toEqual('');
  });

  it('should wrap in button', () => {
    const fixtureApp = TestBed.createComponent(TestAppComponent);
    fixtureApp.detectChanges();
    const button = fixtureApp.nativeElement.querySelector('button');
    expect(button.className).toContain('md-button--none');
  });

  it('should handle click event', () => {
    const fixtureApp = TestBed.createComponent(TestAppComponent);
    const testAppComponent = fixtureApp.componentInstance;
    fixtureApp.detectChanges();
    const button = fixtureApp.nativeElement.querySelector('button');
    button.click();
    fixtureApp.detectChanges();
    expect(testAppComponent.count).toEqual(1);
  });

  it('should render when title is only spaces', () => {
    testComponent.title = '  ';
    testComponent.src = testSrc;
    testComponent.size = 36;
    fixture.detectChanges();
    expect(
      fixture.nativeElement.querySelector('.md-avatar__letter')
    ).not.toBeNull();
  });
});
