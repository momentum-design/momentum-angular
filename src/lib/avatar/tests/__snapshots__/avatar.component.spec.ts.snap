// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarComponent should match snapshot 1`] = `
<md-avatar
  _size={[Function String]}
  _type=""
  alt=""
  ariaLabel=""
  backgroundColor=""
  buttonClassName=""
  click={[Function EventEmitter]}
  clickable="false"
  color=""
  elementRef={[Function ElementRef]}
  failureBadge="false"
  hasNotification="false"
  hideDefaultTooltip="false"
  image="undefined"
  isDecrypting="false"
  isImageErrored="false"
  isImageLoaded="false"
  isOverview="false"
  title={[Function String]}
>
  <span
    class="md-avatar__letter"
    ng-reflect-klass="md-avatar__letter"
    ng-reflect-ng-class="[object Object]"
    ng-reflect-ng-style="[object Object]"
  >
    TN
  </span>
</md-avatar>
`;
